core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java

```diff
@@ -101,166 +101,60 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
-                if (nativeConfig.enableServer) {
-                    log.warn(
-                            "Options dumpProxies and enableServer are both enabled: this will get the proxies dumped in an unknown external working directory");
-                }
-            }
-            if (nativeConfig.nativeImageXmx.isPresent()) {
-                command.add("-J-Xmx" + nativeConfig.nativeImageXmx.get());
-            }
-            List<String> protocols = new ArrayList<>(2);
-            if (nativeConfig.enableHttpUrlHandler) {
-                protocols.add("http");
-            }
-            if (nativeConfig.enableHttpsUrlHandler) {
-                protocols.add("https");
-            }
-            if (nativeConfig.addAllCharsets) {
-                command.add("-H:+AddAllCharsets");
-            } else {
-                command.add("-H:-AddAllCharsets");
-            }
-            if (!protocols.isEmpty()) {
-                command.add("-H:EnableURLProtocols=" + String.join(",", protocols));
-            }
-            if (nativeConfig.enableAllSecurityServices) {
-                command.add("--enable-all-security-services");
-            }
-            if (!noPIE.isEmpty()) {
-                command.add("-H:NativeLinkerOption=" + noPIE);
-            }
-
-            if (!nativeConfig.enableIsolates) {
-                command.add("-H:-SpawnIsolates");
-            }
-            if (!nativeConfig.enableJni) {
-                log.warn("Your application is setting the deprecated 'quarkus.native.enable-jni' configuration key to false."
-                        + " Please consider removing this configuration key as it is ignored (JNI is always enabled) and it"
-                        + " will be removed in a future Quarkus version.");
-            }
-            if (!nativeConfig.enableServer && !SystemUtils.IS_OS_WINDOWS && !graalVMVersion.isMandrel()) {
-                command.add("--no-server");
-            }
-            if (nativeConfig.enableVmInspection) {
-                command.add("-H:+AllowVMInspection");
-            }
-            if (nativeConfig.autoServiceLoaderRegistration) {
-                command.add("-H:+UseServiceLoaderFeature");
-                //When enabling, at least print what exactly is being added:
-                command.add("-H:+TraceServiceLoaderFeature");
+            List<String> buildCommand = getBuildCommand(nativeConfig, nativeImageProperties,
+                    outputDir, runnerJarName, nativeImage, noPIE, isContainerBuild, graalVMVersion,
+                    executableName);
+            buildCommand = maybeRunInContainer(containerCommand, buildCommand);
+            if (isContainerBuild) {
+                log.info(String.join(" ", buildCommand).replace("$", "\\$"));
             } else {
-                command.add("-H:-UseServiceLoaderFeature");
+                log.info(String.join(" ", buildCommand));
```

Comment:
Why is the `log.info()` conditionalized on `isContainerBuild`? Why `...replace("$", "\\$"))` only then? AFAIK, this is shell-escaping the GC policy in use which might be an inner class (`$` separation). I believe it's because `maybeRunInContainer` conditionally already replaces the `$`. It makes the code overly complicated to read.

Reply:
Correct, and it indeed complicates the code. I wanted the log output to be copy-pasteable so. Suggestions on how to improve this are welcome! Do you think keeping the `log.info` outside of the if and only do the replace in there would make things better?

Reply:
Perhaps don't do the `replace()` call in `maybeRunInContainer` and do the escaping of `$` only when you need it.

Reply:
This means that I will need to check before the invocation if this is a container or not and do the replace. It feels like doing it inside maybeRunInContainer is better. What if I move the log in there?

Reply:
sounds good.